ownCloud/Settings/MediaUploadSettingsSection.swift

```diff
@@ -72,6 +170,194 @@ class MediaUploadSettingsSection: SettingsSection {
+				if let convertSwitch = sender as? UISwitch {
+					self?.changeAndRequestPhotoLibraryAccessForOption(optionSwitch: convertSwitch, completion: { (switchState) in
+						self?.userDefaults.instantUploadPhotos = switchState
+						self?.userDefaults.instantUploadPhotosAfter = switchState ? Date() : nil
+
+						if switchState, let locationSelected = self?.uploadLocationSelected, locationSelected == false {
+							self?.showAccountSelectionViewController()
+						} else {
+							self?.postSettingsChangedNotification()
+						}
+
+					})
+				}
+				}, title: "Instant Upload Photos".localized, value: self.userDefaults.instantUploadPhotos)
+
+			instantUploadVideosRow = StaticTableViewRow(switchWithAction: { [weak self] (_, sender) in
+				if let convertSwitch = sender as? UISwitch {
+					self?.changeAndRequestPhotoLibraryAccessForOption(optionSwitch: convertSwitch, completion: { (switchState) in
+						self?.userDefaults.instantUploadVideos = switchState
+						self?.userDefaults.instantUploaVideosAfter = switchState ? Date() : nil
+
+						if switchState, let locationSelected = self?.uploadLocationSelected, locationSelected == false {
+							self?.showAccountSelectionViewController()
+						} else {
+							self?.postSettingsChangedNotification()
+						}
+					})
+				}
+				}, title: "Instant Upload Videos".localized, value: self.userDefaults.instantUploadVideos)
+
+			bookmarkAndPathSelectionRow = StaticTableViewRow(valueRowWithAction: { [weak self] (_, _) in
+				self?.showAccountSelectionViewController()
+				}, title: "Upload Path".localized, value: "", accessoryType: .disclosureIndicator, identifier: MediaUploadSettingsSection.bookmarkAndPathSelectionRowIdentifier)
+
+			self.add(row: instantUploadPhotosRow!)
+			self.add(row: instantUploadVideosRow!)
+
+			updateDynamicUI()
+		}
+	}
+
+	private func getSelectedBookmark() -> OCBookmark? {
+		if let selectedBookmarkUUID = self.userDefaults.instantUploadBookmarkUUID {
+			let bookmarks : [OCBookmark] = OCBookmarkManager.shared.bookmarks as [OCBookmark]
+			return bookmarks.filter({ $0.uuid == selectedBookmarkUUID}).first
+		}
+		return nil
+	}
+
+	private func updateDynamicUI() {
+
+		func updateInstantUploadSwtches() {
+			OnMainThread {
+				self.instantUploadPhotosRow?.value = self.userDefaults.instantUploadPhotos
+				self.instantUploadVideosRow?.value = self.userDefaults.instantUploadVideos
+			}
+		}
+
+		self.remove(rowWithIdentifier: MediaUploadSettingsSection.bookmarkAndPathSelectionRowIdentifier)
+
+		if let bookmark = getSelectedBookmark(), let path = self.userDefaults.instantUploadPath {
+
+			OCCoreManager.shared.requestCore(for: bookmark, setup: { (_, _) in },
+											 completionHandler: { (core, error) in
+												if let core = core, error == nil {
+													core.fetchUpdates(completionHandler: { (_, _) in
+														self.uploadPathTracking = core.trackItem(atPath: path, trackingHandler: { (_, pathItem, _) in
+															if pathItem != nil {
```

Comment:
Please consider the possibility that the `trackingHandler` may be called several times. That doesn't seem to be the case right now. If you only want to handle the first value, the `isInitial` parameter passed to the `trackingHandler` can help with that.

Reply:
@felix-schwarz fixed